runtime/src/iree/base/internal/file_io.c

```diff
@@ -184,6 +206,147 @@ iree_status_t iree_file_read_contents(const char* path,
   return status;
 }
 
+iree_status_t iree_file_map_contents_readonly(
+    const char* path, iree_allocator_t allocator,
+    iree_file_contents_t** out_contents) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+  IREE_ASSERT_ARGUMENT(path);
+  IREE_ASSERT_ARGUMENT(out_contents);
+  *out_contents = NULL;
+
+  iree_file_contents_t* contents = NULL;
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0,
+      iree_allocator_malloc(allocator, sizeof(*contents), (void**)&contents));
+  contents->allocator = allocator;
+
+  iree_status_t status =
+      iree_file_map_contents_readonly_platform(path, contents);
+
+  if (iree_status_is_ok(status)) {
+    *out_contents = contents;
+  } else {
+    iree_file_contents_free(contents);
+  }
+  IREE_TRACE_ZONE_END(z0);
+  return status;
+}
+
+#if defined(IREE_PLATFORM_ANDROID) || defined(IREE_PLATFORM_IOS) || \
+    defined(IREE_PLATFORM_LINUX) || defined(IREE_PLATFORM_MACOS)
+
+#include <sys/mman.h>
+#include <unistd.h>
+
+static iree_status_t iree_file_map_contents_readonly_platform(
+    const char* path, iree_file_contents_t* contents) {
+  // Open file on disk.
+  FILE* file = fopen(path, "rb");
+  if (file == NULL) {
+    return iree_make_status(IREE_STATUS_NOT_FOUND, "failed to open file '%s'",
+                            path);
+  }
+  contents->mapping = file;
+
+  // Query total file size and ensure the file will fit in memory.
+  uint64_t length = 0;
+  IREE_RETURN_IF_ERROR(iree_file_query_length(file, &length));
+  if (length > (uint64_t)IREE_HOST_SIZE_MAX) {
+    return iree_make_status(IREE_STATUS_RESOURCE_EXHAUSTED,
+                            "file size exceeds host pointer size capacity "
+                            "(64-bit file loaded into a 32-bit program)");
+  }
+
+  // Map the memory.
+  void* ptr =
+      mmap(NULL, (size_t)length, PROT_READ, MAP_SHARED, fileno(file), 0);
+  if (ptr == MAP_FAILED) {
+    return iree_make_status(iree_status_code_from_errno(errno), "mmap failed");
+  }
+
+  contents->const_buffer =
+      iree_make_const_byte_span(ptr, (iree_host_size_t)length);
+  return iree_ok_status();
+}
+
+static void iree_file_contents_free_platform(iree_file_contents_t* contents) {
+  if (contents->mapping) {
+    munmap(contents->buffer.data, (size_t)contents->buffer.data_length);
+    fclose((FILE*)contents->mapping);
+  }
+}
+
+#elif defined(IREE_PLATFORM_WINDOWS)
```

Comment:
Could this use DirectStorage in some way? (https://github.com/microsoft/DirectStorage)

Reply:
that's the long term goal but is device-specific and something that'll happen after the module has been loaded (unless the user creates the device first) - my plan is to use file mapping lookup (GetMappedFileNameA, [hacks on linux scanning /proc/self/maps](https://stackoverflow.com/questions/12237712/how-can-i-show-the-size-of-files-in-proc-it-should-not-be-size-zero/12237783#12237783), etc) to optimize the uploads of data from vmfbs separately from how the vmfb is loaded into memory as that'll vary widely based on user.